### DISCUSSION TITLE
Oven Light and Freezer vs Fridge Door

### DISCUSSION THREAD

**Discussion by aperture01** [thoukydides/homebridge-homeconnect#193] [Q&A]
Hi All,

This is a great plugin. Two quick qestions

1. Can I control the Oven light in apple home like any other light? I don't see this as a controller device. 

2. When the door opens on the fridge it says "Freezer Fridge" is there a way to differentiate between the two "doors". In the home connect app it seems to tell the difference between the two. Can the plugin do the same thing or is this a limitation of the API?

**Comment by thoukydides** [comment]
> 1. Can I control the Oven light in apple home like any other light? I don't see this as a controller device.

According to the Home Connect API documentation, the only lights that can be controlled are:
* [Ambient Light](https://api-docs.home-connect.com/settings/?#ambient-light) on Hood and Dishwasher appliances.
* [Functional Light](https://api-docs.home-connect.com/settings/?#functional-light) on Hood appliances.

However, the additional checking that I added to the plugin recently for unrecognised values returned by the API revealed that the documentation is incomplete. In particular, there are two additional lights (#189) that can be controlled:
* Refrigeration Internal
* Refrigeration External

Those are supported by some FridgeFreezer appliances, and presumably also some Freezer, Refrigerator, and WineCooler appliances. However, given their names, they are very unlikely to be supported by any other appliance types.

I have not seen any evidence of Oven lights being supported... but if there is it will be revealed in the log file for your Homebridge installation. Look for `Unrecognised Setting in Home Connect API` messages.

> 2. When the door opens on the fridge it says "Freezer Fridge" is there a way to differentiate between the two "doors". In the home connect app it seems to tell the difference between the two. Can the plugin do the same thing or is this a limitation of the API?

The Home Connect API originally only supported a single door status per appliance (although it provided two separate door alarm events for FridgeFreezer appliances).

Separate states for each door were [added to the API on 22nd August 2023](https://developer.home-connect.com/changelog):
* [Bottle Cooler Door](https://api-docs.home-connect.com/states/#door-state-bottle-cooler)
* [Chiller Common Door](https://api-docs.home-connect.com/states/#door-state-chiller)
* [Chiller Door](https://api-docs.home-connect.com/states/#door-state-chiller)
* [Chiller Left Door](https://api-docs.home-connect.com/states/#door-state-chiller)
* [Chiller Right Door](https://api-docs.home-connect.com/states/#door-state-chiller)
* [Flex Compartment Door](https://api-docs.home-connect.com/states/#door-state-flex-compartment)
* [Refrigerator 2 Door](https://api-docs.home-connect.com/states/#door-state-refrigerator)
* [Refrigerator 3 Door](https://api-docs.home-connect.com/states/#door-state-refrigerator)
* [Wine Compartment Door](https://api-docs.home-connect.com/states/#door-state-wine-compartment)

When these were added to the API I considered supporting them, but decided against it for a few reasons:
* These different door types are not supported by the [simulated Home Connect appliances](https://developer.home-connect.com/simulator), and I do not have a Home Connect fridge/freezer, so would be unable to test the functionality.
* Some appliances (like the simulated ones) may not have been updated to support these new door types, so it would be necessary to detect them and revert to the current functionality.
* It would be necessary to treat those doors differently to the doors on Oven, Dishwasher, Washer, Dryer, and WasherDryer appliances.
* Migrating from an older version of the plugin would need the old single HomeKit Door service to be removed and new separate Door services added. That would have been quite messy.
* The plugin did not provide configuration options to disable unwanted features, and quite a few users already disliked the single door per appliance, so would probably object to multiple.

The last two of those issues have been addressed by my recent rewrites of the plugin. However, the first three still apply.

**Comment by thoukydides** [comment]
An additional issue with the doors is that the plugin supports the [Open Door](https://api-docs.home-connect.com/commands/?#open-door) command that can command Fridge, Freezer, and FridgeFreezer doors to open. However, the API documentation does not list any equivalents for controlling separate doors... so that functionality would be lost if the plugin replaces the single door with multiple.

I suppose I could provide the option to still create the single Door service so that the user can choose which to use. Making the current single door the default behaviour would also reduce problems when people upgrade the plugin - it would not lose any configuration within the Home app or any scenes/automation that has been setup.

**Comment by thoukydides** [comment]
Looking at the door capabilities... the Freezer and WineCooler appliances only support a single door each so there doesn't seem much point adding the extra doors to those. Hence, only FridgeFreezer and Refrigerator appliances would be affected.

**Comment by thoukydides** [comment]
Please try [v0.36.0](https://github.com/thoukydides/homebridge-homeconnect/releases/tag/v0.36.0).

As discussed above, the individual doors are disabled by default, so need to be manually enabled in the Homebridge UI plugin settings or the `config.json` file.

**Comment by aperture01** [comment]
Oh wow. Thanks for the update! Will give this a try when I get back to the house and let you know. Out for the holidays at the moment. So weird there is no light access for the oven in the API. Seems like an obvious item to want to control. I’ll check the log and see what it spits out when I enable it via homeconnect. One last quick question does the plugin require homebridge 1.7 now? I have an older install with HOOBS that does not work. It says “readablestream is not defined” when it attempts to start the plugin. Figured it was due to this. They haven’t updated in awhile and this occurs on anything past 0.30.2. Thinking I may just switch that to vanilla homebridge. But not sure if it’s just an oversight. Thanks! This is a great plugin. Happy to assist any way I can. On Dec 24, 2023, at 03:08, Alexander Thoukydides ***@***.***> wrote:﻿
Please try v0.36.0.
As discussed above, the individual doors are disabled by default, so need to be manually enabled in the Homebridge UI plugin settings or the config.json file.

**Comment by thoukydides** [comment]
> One last quick question does the plugin require homebridge 1.7 now?

The requirements are detailed in the `package.json` file.

Homebridge 1.4.0 or later should be fine, although I am only testing with 1.7.0. There might be issues with earlier versions, but I don't think that's the case.

> It says “readablestream is not defined” when it attempts to start the plugin. 

Also reported today as issue #195.

This plugin requires either Node.js 18 LTS or Node.js 20 LTS. These are the only versions that Homebridge verified plugins are required to support. End-of-life for Node.js 16 was 11th September 2023. The significant rewrite in v0.31.0 of this plugin removed the backwards compatibility that I previously included for Node.js 16, and this is almost certainly the reason for the `ReadableStream` requirement.

> I have an older install with HOOBS that does not work.

This is a Homebridge plugin, not a HOOBS plugin. I do not support use of HOOBS, and have heard of various issues using this plugin with HOOBS.

My attitude is that if this plugin works with Homebridge but not with HOOBS, then that is an issue for the HOOBS developers, not me.

**Comment by aperture01** [comment]
Fair enough. Makes sense to me. Most certainly is because they haven’t updated their code. Folks using hoobs are in an uproar. Seems they haven’t been keeping pace with things. Will give 0.36 a whirl when I get home. Appreciate the quick response!On Dec 25, 2023, at 18:38, Alexander Thoukydides ***@***.***> wrote:﻿

One last quick question does the plugin require homebridge 1.7 now?

The requirements are detailed in the package.json file.
Homebridge 1.4.0 or later should be fine, although I am only testing with 1.7.0. There might be issues with earlier versions, but I don't think that's the case.

It says “readablestream is not defined” when it attempts to start the plugin.

Also reported today as issue #195.
This plugin requires either Node.js 18 LTS or Node.js 20 LTS. These are the only versions that Homebridge verified plugins are required to support. End-of-life for Node.js 16 was 11th September 2023. The significant rewrite in v0.31.0 of this plugin removed the backwards compatibility that I previously included for Node.js 16, and this is almost certainly the reason for the ReadableStream requirement.

I have an older install with HOOBS that does not work.

This is a Homebridge plugin, not a HOOBS plugin. I do not support use of HOOBS, and have heard of various issues using this plugin with HOOBS.
My attitude is that if this plugin works with Homebridge but not with HOOBS, then that is an issue for the HOOBS developers, not me.

**Comment by aperture01** [comment]
So just wanted to let you know that i’ve completely wiped my hoobs config and rebuilt my environment in Homebridge (couldn’t do a direct upgrade). And this now works! To be fair its a better experience now thats it's working. Pain in the butt initially but its working. The setup for the plugin is much more straightforward with the homebridge GUI. I selected the option for the fridge and freezer “doors” in the config and it works. 

Just as a note it looks like I can use all three. So I guess I could technically still have one that tracks both or separate it into the two, right?

Thanks for implementing this so quickly. I wish there was a way to turn on the oven light with the API though. Seems like an obvious one, not sure why it's not there. 

-Brandon

**Comment by thoukydides** [comment]
Thanks for the update. Homebridge and Homebridge UI have improved significantly over the last few years. Good to hear that you have it up and running.

Each of the services you can enable/disable in the plugin settings operate completely independently. It doesn't make any different to their functionality what combination is enabled. So, yes, in the case of the doors you can have both the combined door and the individual doors if that is useful to you.

There are quite a lot of appliance features not currently exposed via the public Home Connect API, plus a significant number that are available but not documented - look for the `(undocumented)` comments in [`api-value-types.ts`](https://github.com/thoukydides/homebridge-homeconnect/blob/master/src/api-value-types.ts). Many of the missing features are available via the Home Connect app, and a few of them via IFTTT, neither of which use the public API. The public API is being gradually extended over time, as you can see from its changelog: https://developer.home-connect.com/changelog

The Home Connect development team do listen to requests for additional functionality, so it is worth contacting them: https://developer.home-connect.com/support/contact
